articles/mysql-query-performance-optimization-tips/index.md

```diff
@@ -0,0 +1,180 @@
+- Normalize Tables: Normalizing keeps all data non-redundant. The database that is in this state is called 3NF (third normal form). The 3NF ensures that lengthy values such as names, addresses, categories, and contact details are not repeated. Instead, they are represented as IDs across multiple smaller tables. For more details on database normalization, refer [here](https://www.guru99.com/database-normalization.html).
+- Use the Most Appropriate Data Types: There are more than 20 different data types in MySQL designed for different uses.  Some of the data types include Timestamp, DateTime, Integer, ENUM, Float, Double, Char, LongText, and Text. Tables should be designed to minimize space used on a disk. Tables that occupy less disk space results in smaller indexes that can be processed in a shorter duration. For example, if a table will host less than 100 records, you should use the TINYINT data type for the unique id as it takes less space than INT.
+- Avoid Null Values. Declare columns to be NOT NULL where possible. This enables better use of indexes. NULL values increase the processing power needed for testing whether each value is NULL, making SQL operations slower.
+
+#### Use Indexes
+Think of records as content in a book. If you want to learn on a particular subtopic, you would go to the index pages, look for the subtopic you want, then get the page where the subtopic is. Indexes work the same way. They are used to find rows with specific column values much faster. Without using an index, MySQL must begin with searching in the first row and go through the whole table to find the required records. Tables with a huge amount of data are more costly to query.
+
+With the use of an index, MySQL can faster determine the position to seek in the middle of the data file. This is done without going through all the rows and is much faster than reading every row sequentially. Refer to [MySQL developer](https://dev.mysql.com/doc/refman/8.0/en/mysql-indexes.html) guide for more information on indexes.
+
+You can create a single-column or multiple column indexes, as shown below, respectively.
+```sql
+CREATE INDEX tablename_columnname_idx ON tablename (columnname);
+```
+```sql
+CREATE INDEX tablename_column1name_column2name_idx ON tablename (column1name, column2name);
+```
+
+#### Use Wildcards at the End of a Phrase
+In MySQL, wildcards are used in conjunction with the LIKE operator and NOT LIKE operator. They are used to search for data matching some search criteria. You can learn more about wildcards [here](https://www.guru99.com/wildcards.html).
+
+Wildcards result in the most expansive scan when searching for data, which is very inefficient. Leading wildcards are the most inefficient, especially when combined with ending wild cards. In such a case, MySQL has to search all the records for a match. Thus you should avoid leading wild cards. See the queries bellow, one using a leading wildcard and another one using ending wildcards.
```

Comment:
"...one using a leading wildcard and another one using ending wildcards." should be "...one using a leading wildcard and another one using an ending wildcard."